### PR TITLE
Prepare for identitysandbox

### PR DESCRIPTION
**Pull request by blacktm**
⚠️ **Don't merge this!** We're exploring the move to identitysandbox.gov and how to address other agency needs. Here's what we were able to do in this experiment:

- [x] Use the [U.S. Web Design System](https://designsystem.digital.gov) to make the visual experience more inline with the [dev docs](https://developers.login.gov), and make the front-end much easier to maintain
- [x] Replace SAML with OIDC, which makes it much easier for the team to support, and vastly simplifies and modernizes our user management and access
- [x] Admins can create users from email, no longer requiring agencies to have to log in first
- [x] Change from SAML specification language to more general verbiage (like "apps" instead of "service providers")
- [x] Update the list of environments and apps we monitor
- [x] Add ability for users to edit the issuer (turns out forcing issuer structure was a blocker)
- [x] Upgrade to Rails 5
- [x] Vastly improve the usability of forms for creating and editing test apps (redirect URIs, label names and field descriptions)
- [x] Remove the expectation of basic auth

### CODE REVIEW

File: .ruby-version [hunk 1]
```diff
@@ -1 +1 @@
-2.3
+2.5.0
```
Comment (brodygov) [inline .ruby-version]: Currently Login.gov servers run Ruby 2.3.x, and we don't have an easy mechanism to run different applications on different ruby versions.

Comment (blacktm) [reply .ruby-version]: Ah, dang it, I'll drop it down.

**Comment by blacktm** [comment]
Note, this is nearly a rewrite of the app. The important CI tests are passing, but still working through Code Climate, which we can address later. I don't have access to Snyk, so I can't see what that reported issue is.

**Comment by brodygov** [comment]
This fails to deploy in a real environment:

```
---- Begin output of /opt/ruby_build/builds/2.3.6/bin/bundle exec rake assets:precompile ----
STDOUT: 
STDERR: rake aborted!
Figaro::MissingKeys: Missing required configuration keys: ["dashboard_certificate", "dashboard_private_key", "dashboard_private_key_password"]
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/figaro-1.1.1/lib/figaro.rb:28:in `require_keys'
/srv/dashboard/releases/20180215183926/config/initializers/figaro.rb:1:in `<top (required)>'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:286:in `load'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:286:in `block in load'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:258:in `load_dependency'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:286:in `load'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/engine.rb:655:in `block in load_config_initializer'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/notifications.rb:168:in `instrument'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/engine.rb:654:in `load_config_initializer'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/engine.rb:612:in `block (2 levels) in <class:Engine>'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/engine.rb:611:in `each'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/engine.rb:611:in `block in <class:Engine>'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/initializable.rb:30:in `instance_exec'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/initializable.rb:30:in `run'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/initializable.rb:59:in `block in run_initializers'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/initializable.rb:48:in `each'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/initializable.rb:48:in `tsort_each_child'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/initializable.rb:58:in `run_initializers'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/application.rb:353:in `initialize!'
/srv/dashboard/releases/20180215183926/config/environment.rb:2:in `<top (required)>'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:292:in `require'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:292:in `block in require'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:258:in `load_dependency'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/activesupport-5.1.5/lib/active_support/dependencies.rb:292:in `require'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/application.rb:329:in `require_environment!'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/railties-5.1.5/lib/rails/application.rb:445:in `block in run_tasks_blocks'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/sprockets-rails-3.2.1/lib/sprockets/rails/task.rb:62:in `block (2 levels) in define'
/srv/dashboard/shared/bundle/ruby/2.3.0/gems/rake-12.3.0/exe/rake:27:in `<top (required)>'
/opt/ruby_build/builds/2.3.6/bin/bundle:22:in `load'
/opt/ruby_build/builds/2.3.6/bin/bundle:22:in `<main>'
Tasks: TOP => environment
(See full trace by running task with --trace)
---- End output of /opt/ruby_build/builds/2.3.6/bin/bundle exec rake assets:precompile ----
Ran /opt/ruby_build/builds/2.3.6/bin/bundle exec rake assets:precompile returned 1
[2018-02-15T18:40:13+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by blacktm** [comment]
@brodygov Ah yes, would it be possible to change the Figaro key names like so:

`saml_sp_certificate` to `dashboard_certificate`
`saml_sp_private_key` to `dashboard_private_key`
`saml_sp_private_key_password` to `dashboard_private_key_password`

**Comment by brodygov** [comment]
Pull requests welcome. You can find the dashboard config here:

**Comment by blacktm** [comment]
@brodygov Oh cool, haven't looked around in there yet. Looks like there's a few things I need to eventually clean up, but for now I'll just use the existing names. Hopefully it will deploy now 🤞

**Comment by monfresh** [comment]
This is really hard to review with so many changes that are not really related to each other. Ideally, we would adhere to our development practices in all our repos, which means submitting small PRs that deal with one thing at a time. It's also very helpful to be able to look back at a repo's history and see how and why a particular change was made. Upgrading Rails is an especially helpful commit to look at as a checklist for other repos that need to be upgraded.

Also, when upgrading Rails, it is recommended to do it one minor version at a time. For example, from 4.2.x to the latest 5.0.x, then from 5.0.x to the latest 5.1.x. See http://guides.rubyonrails.org/upgrading_ruby_on_rails.html

**Comment by monfresh** [comment]
Gotcha. Sounds like a plan. Thanks for the explanation.

**Comment by mark-from-usds** [comment]
For what it's worth I tried deploying this in my toy environment, and it failed on compiling or installing the ffi gem. I notice you change the ffi version so maybe that's it. Or maybe it needs the ffi C libraries installed. Unfortunately I don't really have time to debug it further.

This looks to me like the cause but there is a long configure/make log I can send you:
conftest.c:3:17: fatal error: ffi.h: No such file or directory

**Comment by blacktm** [comment]
We're slicing off these features in new PRs, closing.

**Pull request closed without merging**